### ISSUE TITLE
Find References not working

### ISSUE THREAD

**Issue by gombosg** [scalameta/metals#3379]
**Describe the bug**

**To Reproduce** Steps to reproduce the behavior:

1. Import project
2. Try a "find references" feature
3. No references found anywhere.

**Expected behavior**

Find references is working!

**Screenshots**

Run doctor says it's all right:
![image](https://user-images.githubusercontent.com/23086060/142980887-a413070f-a9b8-40b9-818e-0c3764d93038.png)
![image](https://user-images.githubusercontent.com/23086060/142981506-413cf14a-db43-481e-ba67-6d668048d696.png)

**Installation:**

- Operating system: Fedora 35
- VSCode version: 1.62.3 (bug is present with insiders, too)
- VSCode extension version: 1.10.15
- Metals version: 0.10.9

**Additional context**

[metals.log](https://github.com/scalameta/metals-vscode/files/7586189/metals.log)

...I can also send SemanticDB files if that helps.

Happens in VSCode main and insiders, tried re-importing, cleaning all data etc. it just won't work. I'd so much use Metals for work as it's blazing fast for code navigation, but this is kind of a deal breaker.
Other smaller projects are fine but I don't know what's the key to reproducing this issue in this project.

**Search terms**

find all references

**Comment by dos65** [comment]
@tgodzik I think I've seen it at least two times but was unable to reproduce. After reloading window everything started working well.

**Comment by kpodsiad** [comment]
> This is the culprit after git bisect: 2422494


@gombosg I'll add more logging points in changes coming from the commit you found and also an option that enables debug output. Then we should have more information about what's going on.

**Comment by gombosg** [comment]
@kpodsiad  
Thanks! It's surely that commit because reverting it on top of `main` helps the issue for me.
You don't necessarily have to release a debugging patch, if that helps you can just add some debugging to a branch and I can compile that and get back with some logs.

**Comment by kpodsiad** [comment]
@gombosg @yaoqiang-se I just merged https://github.com/scalameta/metals/pull/3388. Now 'no build target message' is more verbose and prints file for which presentation compiler isn't loaded. 
Could you use metals snapshot version published from that merge and send metals' output? Maybe it'll point us a direction what's not working.

**Comment by kpodsiad** [comment]
Actually, you could add `-Dswoval.log.level=debug` to `Metals: Server Properties` and then watch `metals.log` using for example `tail -f -n 300 .metals/metals.log` (I don't know what will be the windows equivalent of this)

**Comment by gombosg** [comment]
Thanks for taking this into your hands! I added the debug log argument and compiled the patched version.
What is your email @kpodsiad? All file names in the project are in the log file and this is a closed-source project.

This is the extended no build target error BTW:
```
INFO  no build target found for /home/gombosg/prog/ps/process-street/.metals/.tmp/Main-2313172495481126422.scala. Using presentation compiler with project's scala-library version: 2.13.4[0m
```

**Comment by kpodsiad** [comment]
@gombosg you can use kpodsiadlo@virtuslab.com

> This is the extended no build target error BTW:
> 
> ```
> INFO  no build target found for /home/gombosg/prog/ps/process-street/.metals/.tmp/Main-2313172495481126422.scala. Using presentation compiler with project's scala-library version: 2.13.4�[0m
> ```

I have similar output, I'll investigate what this mysterious file contains, but it seems unrelated to the main problem.

**Comment by uncle-betty** [comment]
After @gombosg narrowed it down, I was able to come up with this change, which fixes the issue for me:

```
diff --git a/metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala b/metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
index 64d4bba68..2bf3c4b12 100644
--- a/metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
+++ b/metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala
@@ -150,9 +150,9 @@ object FileWatcher {
       // and run some tests with `-Dswoval.log.level=debug` to see which files are registered
       // and which file events are received.
       val directoriesToCreate =
-        filesToWatch.sourceDirectories ++ filesToWatch.sourceFiles.map(
-          _.getParent()
-        )
+        filesToWatch.semanticdDirectories ++
+        filesToWatch.sourceDirectories ++
+        filesToWatch.sourceFiles.map(_.getParent())
 
       val createdDirectories = directoriesToCreate.flatMap(path =>
         AbsolutePath(path).createAndGetDirectories()
```

According to a source code comment, file watching has issues when paths to non-existing directories are registered to be watched. As a workaround, the code creates any missing directories, then does the registration, and afterwards deletes the created directories again.

It was a bit of a Hail Mary when I simply added the directories that contain the `.semanticdb` files to the list of directories to be created when non-existent. Bloop seems to use a randomly named directory under `.bloop/root/bloop-bsp-clients-classes` for the `.semanticdb` files. And when the registration code runs, this directory hasn't yet been created. So, I figured it would be plausible, if this solved the issue for me. And it did.

**Comment by kpodsiad** [comment]
@uncle-betty do you have problems with closed source project? It would be nice to have a open source project when we can reproduce this issue :/

**Comment by uncle-betty** [comment]
Unfortunately, my project is closed source. But I just managed to create a minimal working example: [mwe.zip](https://github.com/scalameta/metals/files/7766401/mwe.zip)

Here's how I am able to reproduce the problem:

  * Ensure that no Java processes are running, i.e., no Bloop, no SBT, nothing.
  * `unzip mwe.zip`
  * `cd mwe`
  * `vim app/test/Test.scala` - this starts Metals 0.10.9.
  * Import the build when prompted.
  * Wait for the import to finish. (Watch `.metals/metals.log`.)
  * Try to find references for `foo()`.

It seems that it's not enough for the watched directory to not exist during registration. I also needed to enable the Play Framework's plugin in order to reproduce the problem. So, maybe some weird interaction between Play, Bloop, and Metals?

Without the Play Framework's plugin, Metals still registers a non-existent directory for the `.semanticdb` files. But in this case everything works as expected: once the directory has subsequently been created and is being populated with `.semanticdb` files, the corresponding watch events fire.

So, non-existent directories by themselves don't seem to be the problem. They seem to only become a problem when the plugin is enabled. How odd. (On the other hand - I guess that the Play Framework's plugin also watches files... So at least a weird interaction between different watchers is plausible.)

Also note that this problem might not exist on macOS. So it might not be reproducible on a Mac. On macOS the code seems to just register and watch the project root directory instead of multiple individual (and potentially empty) subdirectories of the project.

**Comment by gombosg** [comment]
Very nice  @uncle-betty ! Unfortunately #3395 doesn't seem to work for me. But your repro works, as in find references is not working for `foo()` and also it works after reverting 2422494.

**Comment by uncle-betty** [comment]
Cool, thanks for trying, @gombosg.

Well, it looks like me taking wild guesses won't get us anywhere. :) So, maybe collect some more data on the issue? I created a little experiment that allows us to better understand the behavior of the `file-tree-views` library that Metals's change detection is based on: [repro.zip](https://github.com/scalameta/metals/files/7775763/repro.zip)

Here's how I use it. First, unpack it and build the Scala part:

  * `unzip repro.zip`
  * `cd repro`
  * `sbt assembly`

This gives me `target/scala-2.13/repro-assembly-0.0.jar`.

Then build the C part, which is a shared library that I use to intercept the JVM's inotify-related calls to glibc:

  * `cd inotify`
  * `make`
  * `cd ..`

This gives me `inotify/inotify.so`. Note that this is Linux-specific.

Then run the Scala code with interception enabled:

  * `./run.sh`

The `run.sh` script uses `LD_PRELOAD` to load `inotify.so` into the process subsequently created by `java -jar ...`.

`inotify.so` logs all inotify-related glibc calls to `/tmp/inotify.log.<pid>`, where `<pid>` is the PID of the JVM process. If there isn't any other JVM process running, do the following to watch the correct log:

  * `tail -f /tmp/inotify.log.$(pidof java)`

Alright, let's take a look at the Scala code. All it does is watch a single directory, `A/B/C`. Let's start it with an existing directory `A`:

  * `mkdir A`
  * `./run.sh`

The log file now looks like this:

```
initializing
inotify_init: fd = 6
inotify_init: fd = 10
inotify_add_watch: fd = 10, path = /home/tlo/repro/A, mask = IN_MODIFY:IN_ATTRIB:IN_MOVED_FROM:IN_MOVED_TO:IN_CREATE:IN_DELETE, wd = 1
```

The `file-tree-views` library knows it has to watch `A/B/C`, it has found `A` to exist and it now watches `A` in case `B` gets created. Let's create `B` in a different shell window:

  * `mkdir A/B`

Now this gets logged:

```
read: fd = 10, wd = 1, mask = IN_CREATE:IN_ISDIR, cookie = 0, name = B
inotify_add_watch: fd = 10, path = /home/tlo/repro/A/B, mask = IN_MODIFY:IN_ATTRIB:IN_MOVED_FROM:IN_MOVED_TO:IN_CREATE:IN_DELETE, wd = 2
```

Great! The library noticed creation of `B` inside `A` and now also watches `A/B`, just in case we create `C` inside `A/B`. As expected.

Let's try removing `B` again:

  * `rmdir A/B`

This gets logged:

```
read: fd = 10, wd = 1, mask = IN_DELETE:IN_ISDIR, cookie = 0, name = B
read: fd = 10, wd = 2, mask = IN_IGNORED, cookie = 0, name = (none)
```

Great! We are informed that `B` is gone (`IN_DELETE`) and that the inotify mechanism has removed the watch for it (`IN_IGNORED`).

Let's add `B` back:

  * `mkdir A/B`

We now get the same log output we got when we created `B` before. After all, `A` is still being watched. So the library notices creation of `B` and starts watching it:

```
read: fd = 10, wd = 1, mask = IN_CREATE:IN_ISDIR, cookie = 0, name = B
inotify_add_watch: fd = 10, path = /home/tlo/repro/A/B, mask = IN_MODIFY:IN_ATTRIB:IN_MOVED_FROM:IN_MOVED_TO:IN_CREATE:IN_DELETE, wd = 3
```

So far, so good. Everything seems to work as expected. However, let's now remove `B` and then remove `A`:

  * `rmdir A/B`
  * `rmdir A`

This gets logged:

```
read: fd = 10, wd = 1, mask = IN_DELETE:IN_ISDIR, cookie = 0, name = B
read: fd = 10, wd = 3, mask = IN_IGNORED, cookie = 0, name = (none)
read: fd = 10, wd = 1, mask = IN_IGNORED, cookie = 0, name = (none)
```

Almost as expected: `A`'s watcher gets told that `B` is gone (`IN_DELETE`), `B`'s watcher is removed by inotify (`IN_IGNORED`). OK.

But then someting unexpected (at least to me) happens: `A`'s watcher is also removed by inotify. Fair enough. However, there isn't any watcher that's told that `A` is gone! In other words, there isn't any `IN_DELETE` for `A`, because nobody is watching for `A`'s removal. Somebody should be watching `A`'s parent!

The consequence is the following. Let's recreate `A`:

  * `mkdir A`

... and nothing is logged! So, nobody notices that `A` is back. Consequently, creating B also doesn't log anything:

  * `mkdir A/B`

Generally speaking, I think that we have the following issue at our hand:

  * Suppose that we're watching a path `X1/X2/.../Xn/.../Xm`.
  * Suppose that the following part of the path exists at the time of registration with the `file-tree-views` library: `X1/X2/.../Xn`
  * We're fine as long as we only create and remove directories `X(n+1)` through `Xm`.
  * If we remove any of the directories `X1` through `Xn`, change tracking stops.

Or, more simply speaking: We break change tracking when we remove a directory above what was there at registration time.

My guess would be that this might be a bug in `file-tree-views`, no? The fix might be to watch `X1`, `X1/X2`, ..., `X1/X2/.../Xn`, i.e., to watch all prefixes of `X1/X2/.../Xn` in addition to `X1/X2/.../Xn`.

I'm not sure that this is what's causing the Metals trouble. But it certainly seems like `file-tree-views` is not always robust in scenarios where watched directories are created and removed.

Next, I'll apply `inotify.so` to Metals instead of my toy Scala project. I'll make `inotify.so` additionally intercept calls to `mkdir()` and `rmdir()`, so that I better understand which directories Metals creates and removes in which order and how this interleaves with any inotify-related calls Metals makes.

**Comment by uncle-betty** [comment]
In case somebody else wants to poke around in Metals, here's the version of `inotify.so` that I use for Metals: [inotify.zip](https://github.com/scalameta/metals/files/7776321/inotify.zip)

From what I've seen so far, it does not look like the above unexpected behavior of `file-tree-views` has anything to do with the Metals issue, though. In my MWE, the directories on the path of the `.semanticdb` files simply get created. No directory is ever deleted, so this `file-tree-views` behavior wouldn't apply:

  * `sbt bloopInstall` creates `.bloop/root`.
  * Then the Bloop server creates `.bloop/root/bloop-bsp-clients-classes/classes-Metals-PqKwH4KjSKiHWt4cyZkjbA==`.
  * Then Metals registers the watches.
  * Then the Bloop server creates `.bloop/root/bloop-bsp-clients-classes/classes-Metals-PqKwH4KjSKiHWt4cyZkjbA==/META-INF/semanticdb/app/test`.

How to use `inotify.zip`:

  * `unzip inotify.zip`
  * `cd inotify`
  * `make`
  * `cp inotify.so /tmp`
  * `cp metals.sh /tmp`
  * `vim /tmp/metals.sh` to modify the `METALS` variable to point to the Metals executable. (Mine is `/tmp/metals-vim`.)
  * Change the path to Metals in the editor plugin configuration to point to `/tmp/metals.sh` instead of the Metals executable.

So, `metals.sh` is just a wrapper script that invokes the Metals executable with `LD_PRELOAD` set accordingly. As before, any logging goes to `/tmp/inotify.log.<pid>`.

Alright. I'm off to do some more poking.

**Comment by uncle-betty** [comment]
Ah! I think that I might finally be getting somewhere. After reading a bit of `file-tree-views`'s `NioPathWatcher.java`, I went back to my little `repro.zip` experiment. Suppose the following setup:

  * The following directory exists: `A/B2`. `A` is empty other than containing `B2`.
  * Register a watch for `A/B1`.
  * Register a watch for `A/B2/C`.

Here's what happens in `NioPathWatcher.findOrAdd()`:

  * As `A/B1` does not exist, the first watch's path is truncated to the longest prefix of `A/B1` that exists. `A` exists, so the longest existing prefix would be `A`. So there's now a watch for `A` instead of `A/B1`.

  * When the second watch is registered, the code notices that a watch for `A` already exists. As `A` is a prefix of `A/B2/C`, no new watch gets registered. I think that this is a bug.

  * When I now create `A/B2/C`, `file-tree-views` does not notice it, because it did not set up a watch for `A/B2`.

I think that this might explain the symptoms of our Metals problem. Remember @pvid's comment in the Metals code, where he points out that creating source code directories for some reason seems to help with detecting changes in `.semanticdb` files? Creating these directories tends to lengthen the longest existing prefix of paths. So, the likelyhood that the longest existing prefix of one (source code) path is a prefix of another (semanticdb) path - i.e., the likelyhood of triggering the above bug - is reduced.

**Comment by uncle-betty** [comment]
One last thought for today: If my analysis is correct, then I think that a workaround in Metals would be to register the paths reverse-ordered by the length of their longest existing prefix. In my toy example, we'd register `A/B2/C` (longest existing prefix: `A/B2`, i.e., length 2) before `A/B1` (longest existing prefix `A`, i.e., length 1).

Let me see whether already I know enough Scala to try this...

**Comment by uncle-betty** [comment]
@gombosg, could you try out the proposed workaround in #3404? It seems to do the trick for my MWE as well as for my real (closed source) project.

**Comment by uncle-betty** [comment]
Quick update: There is now version 2.1.8 of `file-tree-views`. Building Metals with this version instead of 2.1.7 also fixes the Metals issue for me.

**Comment by gombosg** [comment]
@uncle-betty nice find! It was a direct fix for this issue, see the commit linked above. https://github.com/swoval/swoval/commit/40278472cdeff02be8feebf26eaf7a538a5e9561 As I see, the fix is based on your findings, so it was very useful that you dug into this. I haven't had the patience to do so. 🙂

**Comment by kpodsiad** [comment]
Wow, @uncle-betty. Your work is incredible, there are simply no words to describe what you've done!

**Comment by uncle-betty** [comment]
Thank you very much for your kind words, guys. I'm glad that I could help.

Let's also not forget @eatkins who analyzed the issue in practically no time and released an upgrade of his component within hours. Awesome work! Thank you!

**Comment by gombosg** [comment]
It works great, I tested editing files, adding new files! 🎉

**Comment by ckipp01** [comment]
Closed via #3405 

Thanks again for all your work on this!
